04-spiral_classification.ipynb

```diff
@@ -308,6 +308,78 @@
    "source": [
     "# ! ffmpeg -i {m}-3d-levels/%03d.png -vf \"crop=trunc(iw/2)*2:trunc(ih/2)*2\" -pix_fmt yuv420p {m}-3d-levels.mp4"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Low dimensional embeddings"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model = nn.Sequential(\n",
+    "    nn.Linear(n, d),\n",
+    "    nn.ReLU(),\n",
+    "    nn.Linear(d, 2),  # new layer with output in R^2\n",
+    "    nn.Linear(2, K)\n",
+    ")\n",
+    "model.to(device)\n",
+    "\n",
+    "C = torch.nn.CrossEntropyLoss(reduction='none')\n",
+    "optimiser = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=lambda_l2)\n",
+    "\n",
+    "for t in range(2_000):\n",
+    "    l = model(X)\n",
+    "    F = C(l, y)\n",
+    "    L = F.mean()\n",
+    "    optimiser.zero_grad()\n",
+    "    L.backward()\n",
+    "    optimiser.step()\n",
+    "    overwrite(f'[EPOCH]: {t}, [LOSS]: {L.item():.6f}, [ACCURACY]: {acc(l, y):.3f}')\n"
```

Comment:
Why is there another training here?

Reply:
I re-used the variable name `model`, though it is a different model with an extra linear layer, such that data goes $\mathbb{R}^n \rightarrow \mathbb{R}^d \rightarrow \mathbb{R}^2 \rightarrow \mathbb{R}^K$ instead of $\mathbb{R}^n \rightarrow \mathbb{R}^d \rightarrow \mathbb{R}^K$ in the previous one. 

The added code just trains that new model and plots the internal 2D embeddings from that extra layer. 

Please feel free to close the pull request without merging it if this adds too much code and not much value ^^

Reply:
OMG, $\LaTeX$ on GitHub! I saw the announcement, but I haven't used it yet! Aww! ❤️

I fear that this would be take away the attention from the major points.
There's the linear model.
There's the 'deep model'.
There are two ways of looking at the energy.
This notebook doesn't focus on classification.
Classification is what we use to introduce a bunch of other things.

I don't want to close this PR.
Let me think what we could do.